### ISSUE TITLE
no-unsed-vars incorrectly determines a write only variable

### ISSUE THREAD

**Issue by TristanWright** [eslint/eslint#7706]
**Tell us about your environment**

- **ESLint Version**: 3.7.1
- **Node Version**: 5.10.1
- **npm Version**: 3.8.3

**What parser (default, Babel-ESLint, etc.) are you using?** Babel-ESLint

**Please show your full configuration: What did you do? Please include the actual source code causing the issue.** 

A few locations, [here is a short one](https://github.com/Kitware/vtk-js/blob/586dc33c9c7f8d25d9b93b15d316abfc31008aa0/Sources/Rendering/OpenGL/ImageMapper/index.js#L355-L358): 
```js
let cellOffset = 0;
cellOffset += model.tris.getCABO().createVBO(cells,
  'polys', VTK_REPRESENTATION.SURFACE,
  { points, tcoords, cellOffset });
```

Here is [a much longer example](https://github.com/Kitware/vtk-js/blob/76baee271c78463cb5ddda7ff5c81d83e02f5677/Sources/Rendering/OpenGL/PolyDataMapper/index.js#L1208-L1252) where cellOffset is +='ed several times but still flagged as an unused var.

An even shorter example I ran on http://eslint.org/demo/
```js
var val = 0;

// increments foo from an object and returns its value
function myFunc(obj) {
  obj.foo += 1;
  return obj.foo;
}

// pass val as the value of foo, val is still marked as unused?
val = myFunc({ foo: val });
val += myFunc({ foo: val });
```

**What did you expect to happen?**
Expected `cellOffset` to not be marked as an unused var.

**What actually happened? Please include the actual, raw output from ESLint.**
`cellOffset` is marked as unused under the rule `no-unused-vars`.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the report.

I'm pretty sure this is working as expected. Your code is *writing* to `cellOffset`, but the value of `cellOffset` isn't actually getting used afterwards, so your code would work exactly the same if you didn't write to the variable.

Your code currently looks like this:

```js
let cellOffset = 0;
cellOffset += model.tris.getCABO().createVBO(cells,
  'polys', VTK_REPRESENTATION.SURFACE,
  { points, tcoords, cellOffset });
```

But it would work the same way if it looked like this:

```js
let cellOffset = 0;
model.tris.getCABO().createVBO(cells,
  'polys', VTK_REPRESENTATION.SURFACE,
  { points, tcoords, cellOffset });
```

**Comment by TristanWright** [comment]
I think I understand but isn't the value used in the object passed to the function? 

`{ points, tcoords, cellOffset }`

**Comment by platinumazure** [comment]
@TristanWright Could you please construct a simpler example? It'd be good to see if this is due to destructuring or if this applies to any function call where the "unused" variable is passed as an argument.

For example, do these result in problems?

```js
let foo = 0;
foo += someFunc(foo);

let bar = 0;
bar += someOtherFunc({ bar });
```

**Comment by not-an-aardvark** [comment]
It is used there, but after the function call, you're assigning it a value again (with `cellOffset += foo`). The variable isn't used after it's assigned that value, so the assignment is unnecessary.

I noticed you're using ESLint 3.7.1. In ESLint 3.8.0, the error message for `no-unused-vars` was slightly improved to be more accurate for problems like this.

**Comment by TristanWright** [comment]
@platinumazure the code block you pasted reports `foo` and `bar` are unused. With newer eslint versions too, but I'm convinced now this is the correct behavior.

@not-an-aardvark I see now, one could think of each variable having a 'used'/'unused' state. Passing it as an argument, in an object, or operators etc. it becomes used, assigning to it it becomes unused. Thanks for all your help in clarifying this!
